src/trap_routines.rs

```diff
@@ -0,0 +1,272 @@
+use crate::{
+    error::VMError,
+    hardware::{Memory, Register, Registers},
+    utils::{getchar, stdout_flush, stdout_write, update_flags},
+};
+use std::{
+    io::{Read, Write},
+    num::TryFromIntError,
+};
+
+const NULL: u16 = 0x0000;
+
+pub enum TrapCode {
+    GetC,
+    Out,
+    Puts,
+    In,
+    PutsP,
+    Halt,
+}
+
+impl TryFrom<u16> for TrapCode {
+    type Error = VMError;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        match value {
+            0x20 => Ok(TrapCode::GetC),
+            0x21 => Ok(TrapCode::Out),
+            0x22 => Ok(TrapCode::Puts),
+            0x23 => Ok(TrapCode::In),
+            0x24 => Ok(TrapCode::PutsP),
+            0x25 => Ok(TrapCode::Halt),
+            _ => Err(VMError::Conversion(String::from(
+                "Invalid u16 for TrapCode conversion",
+            ))),
+        }
+    }
+}
+
+/// Read one character from the stdin.
+pub fn get_c(regs: &mut Registers, reader: &mut impl Read) -> Result<(), VMError> {
+    let buffer = getchar(reader)?;
+    let char: u16 = buffer[0].into();
+    regs[Register::R0] = char;
+    update_flags(Register::R0, regs);
+    Ok(())
+}
+
+/// Output a single character.
+pub fn out(regs: &mut Registers, writer: &mut impl Write) -> Result<(), VMError> {
+    let c: u8 = regs[Register::R0]
+        .try_into()
+        .map_err(|e: TryFromIntError| VMError::Conversion(e.to_string()))?;
+    stdout_write(&[c], writer)?;
+    Ok(())
+}
+
+/// Prompt for input character from the user.
+pub fn trap_in(
+    regs: &mut Registers,
+    writer: &mut impl Write,
+    reader: &mut impl Read,
+) -> Result<(), VMError> {
+    print!("Enter a character: ");
+    let buffer = getchar(reader)?;
+    stdout_write(&buffer, writer)?;
+    stdout_flush(writer)?;
+    regs[Register::R0] = buffer[0].into();
+    update_flags(Register::R0, regs);
+    Ok(())
+}
+
+/// Output a null-terminated string. The characters are contained in consecutive memory locations,
```

Comment:
To keep consistency with the documentation of other functions:

```suggestion
/// Writes a null-terminated string into stdout. The characters are contained in consecutive memory locations,
```